content/things/webgpu.md

```diff
@@ -0,0 +1,586 @@
+
+> **Note:** The term “shader” used to confuse me, because you can do _so much more_ than just shading. But in the olden days (late 1980s!), that term was appropriate: It was a small piece of code that ran on the GPU to decide for each pixel what color it should be so that you could _shade_ the objects being rendered, achieving the illusion of lighting and shadows. Nowadays, shaders loosely refer to any program that runs on the GPU.
+
+The GPU now rasterizes the triangles, meaning the GPU figures out which pixels each triangle covers on the screen. Each pixel is then processed by the _fragment shader_, which has access to the pixel coordinates but also the auxillary data to decide which color that pixel should be. When used correctly, amazing 3D graphics can be created with this process.
+
+This system of passing data to a vertex shader, then to a fragment shader and then outputting it directly onto the screen is called a _pipeline_, and in WebGPU you have to explicitly define your pipeline.
+
+### Pipelines
+
+Currently, WebGPU allows you to create two types of pipelines: A Render Pipeline and a Compute Pipeline. As the name suggest, the Render Pipeline renders something, meaning it creates a 2D image. That image needn’t be on screen, but could just be rendered to memory (which is called a Framebuffer). A Compute Pipeline is more generic in that it returns a buffer, which can contain any sort of data. For the remainder of this blog post I’ll focus on Compute Pipelines, as I like to think of Render Pipelines as a specialization/optimization of Compute Pipelines. Now, this is not only historically backwards — the compute pipeline was built as an abstraction over the very purpose-built rendering pipeline — but actually considerably understates that these pipelines are physically different circuits in your GPU. In terms of the API, however, I find my mental model quite good. <mark>In the future, it seems likely that more types of pipelines — maybe a Raytracing Pipeline — get added to WebGPU.</mark>
+
+With WebGPU, a pipeline consists of one (or more) programmable stages, where each stage is defined by a shader and an entry point. A Compute Pipline has a single `compute` stage, while a Render Pipeline would have a `vertex` and a `fragment` stage:
+
+```js
+const module = device.createShaderModule({
+  code: `
+    @stage(compute) @workgroup_size(64)
+    fn main() {
+      // Pointless!
+    }
+  `,
+});
+
+const pipeline = device.createComputePipeline({
+  compute: {
+    module,
+    entryPoint: "main",
+  },
+});
+```
+
+<mark>Should I recommend using `createComputePipelineAsync` instead?</mark>
```

Comment:
Turns out this is slightly in flux, so for the moment simply using `createComputePipeline()` is safe.